File: 2273-find-resultant-array-after-removing-anagrams.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

class FindResultantArrayAfterRemovingAnagramsSolution {

    public String sortedString(String s) {
        char[] charArray = s.toCharArray();
        Arrays.sort(charArray);
        return new String(charArray);
    }

    public List<String> removeAnagrams(String[] words) {
        int n = words.length;
        String prevSorted = sortedString(words[0]);
        int i = 1;
        int idxToAdd = 0;
        List<String> result = new ArrayList<>();
        while (i < n) {
            String currSorted = sortedString(words[i]);
            if (!currSorted.equalsIgnoreCase(prevSorted)) {
                result.add(words[idxToAdd]);
                idxToAdd = i;
            }

            i++;
            prevSorted = currSorted;
        }
        result.add(words[idxToAdd]);

        return result;
    }

    public static void main(String[] args) {
        FindResultantArrayAfterRemovingAnagramsSolution sol = new FindResultantArrayAfterRemovingAnagramsSolution();
        System.out.println("Output: "
                + Arrays.toString(
                        sol.removeAnagrams(new String[] { "abba", "baba", "bbaa", "cd", "cd" }).toArray(new String[0]))
                + ", expected: [\"abba\", \"cd\"]");
        System.out.println("Output: "
                + Arrays.toString(sol.removeAnagrams(new String[] { "a", "b", "c", "d", "e" }).toArray(new String[0]))
                + ", expected: [\"a\",\"b\",\"c\",\"d\",\"e\"]");
    }

}

File: 2273-find-resultant-array-after-removing-anagrams.py
class Solution:
    def removeAnagrams(self, words: list[str]) -> list[str]:
        '''
        We go through each word, starting from index 1.
        We will have a prev_sorted and curr_sorted 
            - If the 2 strings are equal, we mark the curr index as the index to use
            - Else, we add the word pointed to by the curr index into the result array and set the prev_sorted as the curr_sorted 
        '''
        result = []
        prev_sorted = sorted(words[0])
        i = 1
        n = len(words)
        index_to_add = 0
        while i < n:
            curr_sorted = sorted(words[i])
            if prev_sorted != curr_sorted:
                result.append(words[index_to_add])
                index_to_add = i

            i += 1
            prev_sorted = curr_sorted

        result.append(words[index_to_add])

        return result


sol = Solution()
print(
    f'output: {sol.removeAnagrams(["abba", "baba", "bbaa", "cd", "cd"])}, expected: ["abba", "cd"]')
print(
    f'output: {sol.removeAnagrams(["a", "b", "c", "d", "e"])}, expected: ["a","b","c","d","e"]')
